gpt_index/indices/keyword_table/base.py

```diff
@@ -177,6 +177,14 @@ def _delete(self, doc_id: str, **delete_kwargs: Any) -> None:
         for keyword in keywords_to_delete:
             del self._index_struct.table[keyword]
 
+    def _find_matching_hash(self, doc_id: str) -> Optional[str]:
```

Comment:
ah iterating over nodes just to find the reference doc hash seems a bit hacky (and will be slow as nodes increase). 

Here's a proposal: currently the `DocumentStore` object on an index (defined in `gpt_index/docstore.py`) doesn't contain document information, it only contains index_struct information. We actually took out document information in an earlier release to save space, since we were storing the full text per document and that was unnecessary. 

Instead, we could add a `ref_doc_info` field to the DocumentStore, which stores references from doc_id to a dictionary of "metadata" for that document. The first entry in that metadata could be the hash of the document. That way, _find_matching_hash could just look for the hash through a docstore lookup, and could be implemented in the base index class. 

what are your thoughts on this?

Reply:
Ah, I think that would work nicely (and still keeps the docstore small!)

I'll work on that later today 💪